lib/src/components/NxModal/NxModal.tsx

```diff
@@ -27,60 +27,98 @@ const NxModal: FunctionComponent<Props> = ({className, onClose, variant, role, .
-    currentModalCloseHandlers.splice(idx, 1);
-
-    if (!currentModalCloseHandlers.length) {
-      document.removeEventListener('keydown', modalCloseListener);
+  function dialogKeydownListener(evt: KeyboardEvent<HTMLDialogElement>) {
+    if (evt.key === 'Escape' || evt.key === 'Esc') {
+      // HACK for backwards compatibility: it is known that some downstream uses of NxModal do not provide
+      // the onCancel/onClose handler despite it being required. It is also known that these downstream uses
+      // have some logic which globally manages ESC handlers for a variety of things including modals. To keep that
+      // working, only stopPropagation if the onCancel callback is defined. If it isn't defined, tenuously assume
+      // that the ESC handling is implemented externall and do nothing here.
+      if (onCancel) {
+        // prevent visibility of the keydown outside of the modal, so that global ESC listeners on the
+        // document don't pick it up
+        evt.stopPropagation();
+
+        // prevent visibility to manually-registered native event listeners on the document too.
+        // NOTE: this only works on listeners added after this one, which is believed to include any
+        // registered in useEffect calls on components rendered simultaneously with the modal
+        evt.nativeEvent.stopImmediatePropagation();
+
+        if (!hasNativeModalSupport && !evt.defaultPrevented) {
+          // emulate cancel-on-esc behavior in browsers which don't do it natively
+          onCancel(new Event('cancel', { cancelable: true }));
+        }
+      }
     }
-  };
+  }
 
   useEffect(function() {
-    if (!currentModalCloseHandlers.length) {
-      document.addEventListener('keydown', modalCloseListener);
+    /* eslint-disable-next-line @typescript-eslint/no-non-null-assertion */
+    const el = dialogRef.current!,
+
+        // HTML <dialog> elements are supposed to remember what element was focused before they were opened,
+        // and restore focus to that element when they are closed. No browsers appear to implement this currently,
+        // so we do it ourselves
+        previouslyFocusedEl = document.activeElement;
+
+    setDialogRefState(el);
+
+    /*
+     * This will cause the document to become "blocked by the modal dialog"
+     * (https://html.spec.whatwg.org/multipage/interaction.html#blocked-by-a-modal-dialog)
+     * meaning that only the modal and its contents are interactable/focusable.
+     *
+     * Note: not supported in safari, which is why the conditional is here. Once safari gets
+     * it together, we should be able to take advantage of the "top layer" functionality built into
+     * the browser around modals to simplify the NxModal styling around z-index handling
+     */
+    if (hasNativeModalSupport) {
+      el.showModal();
+    }
+    else {
+      // without native support we don't trap focus in the modal, but we can at least start it off there
```

Comment:
Is that "we don't" or "we can't"?

Reply:
> Is that "we don't" or "we can't"?

Don't. It's something that can attempted in JavaScript, more or less successfully, but I always found it hacky and error prone and now I feel comfortable just blaming browsers for not supporting `<dialog>` if anyone points out that we don't do it.